### PR TITLE
charmcraft: perform LXD checks before use (CRAFT-420)

### PR DESCRIPTION
**Pull request by cjp256**
Invoke ensure_lxd_is_ready() interface before using LXD
provider. This consolidates a couple of existing checks
done elsewhere as well as a new one for ensuring the user
has the necessary permissions to use LXD.

Minor simplification of imports used in providers module
to improve and update mocks accordingly.

Example error:
```
$ /snap/bin/charmcraft pack
LXD requires additional permissions.
Please ensure that the user is in the 'lxd' group.
(full execution logs in '/home/test1/snap/charmcraft/common/charmcraft-log-0xgcist7')
```

Signed-off-by: Chris Patterson <chris.patterson@canonical.com>

### CODE REVIEW

File: charmcraft/providers.py [hunk 1]
```diff
@@ -112,26 +106,29 @@ def ensure_provider_is_available() -> None:
...
+        if confirm_with_user(
+            "LXD is required, but not installed. Do you wish to install LXD "
+            "and configure it with the defaults?",
+            default=False,
+        ):
+            try:
+                lxd.install()
+            except lxd.LXDInstallationError as error:
+                raise CommandError(
+                    "Failed to install LXD. Please visit https://snapcraft.io/lxd for "
```
Comment (sergiusens) [inline charmcraft/providers.py:119]: Remove Please, start with Visit -> https://discourse.ubuntu.com/t/tone-of-voice/18883

File: charmcraft/providers.py [hunk 2]
```diff
@@ -112,26 +106,29 @@ def ensure_provider_is_available() -> None:
...
+                    "instructions on how to install the LXD snap for your distribution"
+                ) from error
+        else:
             raise CommandError(
-                "Failed to install LXD. Please visit https://snapcraft.io/lxd for "
+                "LXD is required, but not installed. Please visit https://snapcraft.io/lxd for "
```
Comment (sergiusens) [inline charmcraft/providers.py:124]: same

**Review by sergiusens — CHANGES_REQUESTED** [review]
Looks generally good, tone of voice needs adjusting.

File: charmcraft/providers.py (hunk 1 above)
Comment (cjp256) [reply charmcraft/providers.py:119]: Ah yes, I questioned it when I wrote it way back when and never did reconsider. :thinking:  I think I just feel compelled to be polite :rofl:  will fix (but note that this was just an indent change, not new text).

Comment (cjp256) [reply charmcraft/providers.py]: done!

File: charmcraft/providers.py (hunk 2 above)
Comment (cjp256) [reply charmcraft/providers.py]: done!

**Review by sergiusens — APPROVED** [review]

File: charmcraft/providers.py [hunk 3]
```diff
@@ -112,26 +106,29 @@ def ensure_provider_is_available() -> None:
...
-        "and configure it with the defaults?",
-        default=False,
-    ):
-        try:
-            lxd_installer.install()
-        except LXDInstallationError as error:
+    if not is_provider_available():
+        if confirm_with_user(
+            "LXD is required, but not installed. Do you wish to install LXD "
+            "and configure it with the defaults?",
```
Comment (facundobatista) [inline charmcraft/providers.py:112]: Not worth for this branch, but for the future: for readability in these cases please write the long message outside the `if`. Thanks!

**Review by facundobatista — APPROVED** [review]
Looks great!

**Pull request merged**